### ISSUE TITLE
training run: Lora Ichigo Qwen2.5 32B

### ISSUE THREAD

**Issue by bachvudinh** [janhq/ichigo#134]
## Goal
- Evaluate Qwen2.5's capabilities, which surpasses LLaMA 3.1 across all English benchmarks and demonstrates particularly strong performance in understanding Asian languages like Vietnamese and Singlish.
- Assess the effectiveness of LoRA training in teaching the model to recognize and process sound tokens.

## Methodology
- Change the base model of Ichigo from Llama 3.1 to Qwen2.5 32B 
- Due to model size constraints, we employed **LoRA adapters across all linear layers** for both continued pretraining and supervised fine-tuning steps, while **fully fine-tuning the embedding and LM head layers** to accommodate 513 new sound tokens. Following Qwen's methodology, we integrated control tokens into the embedding layer without tokenizer modification (as discussed in Qwen2.5 Issue #29) which, according to Qwen Authors, aim at optimizing training performance(NVIDIA's matrix multiplication guidelines: https://docs.nvidia.com/deeplearning/performance/dl-performance-matrix-multiplication/index.html#), we adjusted the embedding layer and LM head dimensions by a factor of 128, resulting in a final embedding size of 152,192 tokens.
  ```
  ["q_proj", "k_proj", "v_proj", "o_proj", "gate_proj", "down_proj", "up_proj"]
  ```
- For pretrain data, I used only English subset of [Pretrain Data](https://huggingface.co/datasets/homebrewltd/raw-speech-whispervq-v2) and SFT on [v0.4 dataset](https://huggingface.co/datasets/homebrewltd/mixed-instruction-speech-whispervq-v4.) 

## Experiments

| Run ID | Date       | Model Config | Dataset | Sequence Length | Learning Rate | Batch Size | Steps | Loss | Hardware | MMLU | MMLU pro | Notes |
|--------|------------|--------------|---------|-----------|---------------|------------|-------|------|----------|------|-----------|--------|
| exp1-pretrain | 2024-11-23 | [Lora-256-512](https://github.com/janhq/research/blob/main/assets/ichigo-32b/exp1/config/32B_lora_pretrain.yaml) | [Pretrain v0.1](https://huggingface.co/datasets/homebrewltd/raw-speech-whispervq-v1) | 512 | 1.5e-4 | 384  | 6302 | [1.9](https://github.com/janhq/research/blob/main/assets/ichigo-32b/exp1/loss/qwen2.5-32b-ichigo-sft.png) | ~ 100 hours on 8xA6000 | - | - | old dataset |
| exp1-sft | 2024-11-27 | [Lora-256-512](https://github.com/janhq/research/blob/main/assets/ichigo-32b/exp1/config/32B_lora.yaml) | [SFT](https://huggingface.co/datasets/homebrewltd/instruction-speech-whispervq-v2-prompt-transcribe.) | 1024 | 3e-4 | 384  | 2500 | [1.2](https://github.com/janhq/research/blob/main/assets/ichigo-32b/exp1/loss/qwen32b-ichigo-pretrain-v0.1.png) | 60/100 hours on 6xA6000 | - | - | stop early to prepare next run |
| exp2-pretrain | 2024-11-26 | [Lora-256-512](https://github.com/janhq/research/blob/main/assets/ichigo-32b/exp1/config/32B_lora_pretrain.yaml) | [Pretrain v0.2](https://huggingface.co/datasets/homebrewltd/raw-speech-whispervq-v2) | 512 | 1.5e-4 | 512  | 4726 | [1.9](https://github.com/janhq/research/blob/main/assets/ichigo-32b/exp1/loss/qwen32b-ichigo-pretrain-v0.2.png) | ~ 100 hours on 8xA6000 | - | - | new dataset v0.2 |
| exp2-sft | 2024-12-1 | [Lora-256-512](https://github.com/janhq/research/blob/main/assets/ichigo-32b/exp1/config/32B_lora.yaml) | [SFT data](https://huggingface.co/datasets/homebrewltd/mixed-instruction-speech-whispervq-v4) | 4096 | 3e-4 | 256  | 8020 | [Updated soon](https://github.com/janhq/research/blob/main/assets/ichigo-32b/exp1/loss/qwen2.5-32b-ichigo-sft.png) | ~ 350 hours on 8xA6000 | - | - | on-going run |

## Quicklinks
- Old Pretrain data v0.1: https://huggingface.co/datasets/homebrewltd/raw-speech-whispervq-v1.
- Pretrain data v0.2: https://huggingface.co/datasets/homebrewltd/raw-speech-whispervq-v2.
- SFT Data: https://huggingface.co/datasets/homebrewltd/mixed-instruction-speech-whispervq-v4
- Checkpoints:

**Comment by dan-homebrew** [comment]
Running on 8 x A6000 in Taiwan

**Comment by tikikun** [comment]
<img width="752" alt="Screenshot 2024-11-29 at 13 06 20" src="https://github.com/user-attachments/assets/e02bcb16-24aa-44db-a548-bab8c9a8bf6e">

**Comment by bachvudinh** [comment]
![llm_training_loss](https://github.com/user-attachments/assets/a0047e89-1fee-4bfd-8a26-9a70d06fef62)

**Comment by dan-homebrew** [comment]
How long would it take on a H100?

**Comment by bachvudinh** [comment]
- We are preparing to release version 0.1 of Lora Ichigo Qwen 2.5 32B, which was trained using an earlier WhisperVQ checkpoint. iam doing some evaluation on some benchmarks like mmli, mmlu pro,... 
- We finished the Pretraining on the [pretrain data](https://huggingface.co/datasets/homebrewltd/raw-speech-whispervq-v2.) on 8A6000 
- Fine-tuning phase is currently ongoing:
  + Using the mixed instruction speech WhisperVQ v4 dataset
  + Timeline has exceeded initial estimates
  + Current projected duration: approximately 350-400 hours
  + Trying to find the optimal learning rate
---> we can't do it on the 8xA6000. 
cc @dan-homebrew @tikikun

**Comment by bachvudinh** [comment]
cc @tikikun @dan-homebrew I reported some issues
- Due to the size of the model 32B, the model can't be trained on our internal cluster. Furthermore, the evaluation of the model on some benchmark like MMLU or MMLU pro take a a lot of time using 8xA6000 (~8 hours for a benchmark):
![image](https://github.com/user-attachments/assets/32c3cff8-030a-42b6-b9f0-9e3b786f3f4e)

- Our current checkpoint works out of the box but its trained on the wrong data so it does not have multi-turn ability and ability to refuse to noisy audio. Given these limitations, releasing this version would not deliver the compelling user experience we aim for with Ichigo.
My initiative and conclusion:
- For Ichigo, we can scale up to a larger model by applying adapters to linear layers while fully fine-tuning only the embedding and language model head. Though this approach makes just 8-10% of parameters trainable compared to full fine-tuning, the model's overall size still results in significant forward pass latency. Looking ahead, implementing LoRA for 8B or 16B models would be more feasible for our cluster's computing capacity.
